[crypto] Introduce PSAKeyAllocator

ArekBalysNordic opened the pull request:
The PSAKeyAllocator introduces a new possibility to manage PSA keys depending on the platform.
Now we can still use common PSA API in the Matter stack, but we can inject the platform-specific implementation for storing the persistent keys. Thanks to that we can use PSA ITS as well as custom a crypto driver which is available on vendor-specific platform

- Allowed setting session keystore by platform. If any keystore is not set, assign the default one.

- Moved the PSA key definitions from the CHIPCryptoPALPSA.h file to
the newly created PSAKeyAllocator.

- The new PSAKeyAllocator class allows for the allocation of keys
in secure storage. Users can create their own PSAKeyAllocator
implementation and set it to be used by the Matter stack.

- If the custom implementation is not provided the default one is
used and it works as the legacy solution the mechanism is
about stored keys in the PSA ITS storage.

### Testing

Added a new unit test for the PSAKeyAllocator under the `TestPSAOpKeyStore` test suite (Separate commit in this PR).
The solution is tested within the examples-nrfconnect workflow where the unit tests of the TestPSAOpKeyStore suite are run on the Posix platform.

ArekBalysNordic committed:
[crypto] Introduce PSAKeyAllocator

- Moved the PSA key definitions from CHIPCryptoPALPSA.h file to
the newly created PSAKeyAllocator.

- The new PSAKeyAllocator class allows for the allocation of keys
in secure storage. Users can create their own PSAKeyAllocator
implementation and set it to be used by the Matter stack.

- If the custom implementation is not provided the default one is
used and it works as the legacy solution and the mechanism is
about stored keys in the PSA ITS storage.

Signed-off-by: Arkadiusz Balys <arkadiusz.balys@nordicsemi.no>

bzbarsky-apple commented on src/app/server/Server.h:
```diff
@@ -278,7 +278,10 @@ struct CommonCaseDeviceServerInitParams : public ServerInitParams
-        this->sessionKeystore = &sSessionKeystore;
+        if (this->sessionKeystore == nullptr)
+        {
+            this->sessionKeystore = &sDefaultSessionKeystore;
```
Realistically: someone who wants to inject a different keystore should not be using the CommonCaseInit thing.  This is making them pay the flash cost for DefaultSessionKeystore even though they are not using it, right?

ArekBalysNordic replied:
That's right, thanks for pointing it. I thought that someone would inherit the base DefaultSessionKeystore, and in this case, it must be compiled anyway. But if it comes to injecting a completely new SessionKeystore implementation, the default one will be redundant, but compiled. I will remove this part from this PR.

LuDuda reviewed (approved):
LGTM :+1:

andy31415 commented on src/crypto/PSAKeyAllocator.h:
```diff
@@ -0,0 +1,216 @@
+}
+
+// Initialize the PSAKeyAllocator instance
+inline PSAKeyAllocator * PSAKeyAllocator::sInstance = nullptr;
```
Would it make sense to create a PSAKeyAllocator.cpp and use `inline` less? I am not sure how things will interact if this gets included in more than one file.

ArekBalysNordic replied:
Thanks to C++ 17 (https://en.cppreference.com/w/cpp/language/inline) we can use inline here and avoid creating and maintaining the additional PSAKeyAllocator.cpp file. Inline variables have been tested on most compilers https://en.cppreference.com/w/cpp/compiler_support/17#cpp_inline_variables_201606L and I assume here that it works as it is designed and this instance will be created only once within the Matter translation unit no matter how many times the header is included. Please correct me if I'm wrong with my thinking, maybe I have wrong assumptions regarding C++17 here, or somewhere in Matter configuration there is something that can cause problems with that, and maybe I'm unaware of it.

ArekBalysNordic replied:
@andy31415 What do you think?

andy31415 replied:
My concern is that this feels like "tricky code". If I am required to read and understand the C++ standard in detail (including differences across versions) to ensure this is safe then the code may not be as "obviously correct" as one would like.

Otherwise sure, I am ok with this as long as you explain in a comment why this is safe and also explain why if this does not work as expected we would blow up/have a test that fails. Right now I am wondering if this works only because a single cpp files includes it or not.

ArekBalysNordic replied:
I found that there is no test that covers your concerns. I've created a new one - `TestKeyAllocation` under the `TestPSAOpKeyStore` suite. It checks the whole functionality of the PSAKeyAllocator and the DefaultPSAKeyAllocator as well. This suite is run within the nrfconnect POSIX unit tests workflow. I've removed locally `inline` from there and the build failed as expected due to multiple definitions. It proves that inline works there and now we have at least one test where it should fail if it does not work as expected. I've added the comment there as well to explain my motivation. Of course, this is all correct if we use C++17, but this is my assumption.

@andy31415 Does this clear up your doubts? :)

ArekBalysNordic committed:
[crypto] Added a unit test for PSAKeyAllocator

The unit test:
- Verifies the defaultKeyAllocator instance, key allocation and
if attributes are ont changed.
- Creates a new testing key allocator and verifies if it works
as expected and differently than the default ones.
- Switches back to the defaultKeyAllocator instance and checks
if code works properly.

Signed-off-by: Arkadiusz Balys <arkadiusz.balys@nordicsemi.no>

soares-sergio commented on src/lib/core/CHIPConfig.h:
```diff
@@ -231,6 +231,39 @@
+ *   Base of the PSA key identifier range used by Matter.
+ *
+ * Cryptographic keys stored in the PSA Internal Trusted Storage must have
+ * a user-assigned identifer from the range PSA_KEY_ID_USER_MIN to
```
identifer -> identifier
